### PR TITLE
[DOC-460] Add SW example to understand usecase for sink, source, table and window [master]

### PR DESCRIPTION
**Pull request by ashish-macrometa**
- Add SW example to understand usecase for sink, source, table and window
- Fixes [[DOC-460](https://macrometa.atlassian.net/browse/DOC-460)]

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves existing functionality)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

[DOC-460]: https://macrometa.atlassian.net/browse/DOC-460?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[DOC-460]: https://macrometa.atlassian.net/browse/DOC-460?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by StoyanTc — APPROVED** [review]

File: docs/cep/tutorials/create-stream-app.md [hunk 1]
```diff
@@ -286,6 +286,73 @@ The stream worker will use the Macrometa collections with the default query para
 </tbody>
 </table>
 
+
+## Source, Sink, Table and Window: Understanding with the help of a Stream Worker example
+
+
+In the application below, a Macrometa document based collection plays role of `source` to our SW IntrusionDetectionSystem.
```
Comment (diana-macrometa) [inline docs/cep/tutorials/create-stream-app.md]:
```suggestion
In the application below, a Macrometa document-based collection plays role of `source` to our SW IntrusionDetectionSystem.
```

File: docs/cep/tutorials/create-stream-app.md [hunk 2]
```diff
@@ -286,6 +286,73 @@ The stream worker will use the Macrometa collections with the default query para
...
+Consider InboundTrafficData as an audit collection where client IPs are added as they make requests to the server.
+As a new record is added, an event is emitted by collection streams associated with this collection. This event is consumed by the `source` InboundTraffic.
+
+Now it is quite possible that your application, `IntrusionDetectionSystem` may get burst of events having the same IP, and you need to gauge
```
Comment (diana-macrometa) [inline docs/cep/tutorials/create-stream-app.md]:
```suggestion
It is quite possible that your application, `IntrusionDetectionSystem`, might get burst of events having the same IP, and you need to gauge  
```

File: docs/cep/tutorials/create-stream-app.md [hunk 3]
```diff
@@ -286,6 +286,73 @@ The stream worker will use the Macrometa collections with the default query para
...
+
+In the application below, a Macrometa document based collection plays role of `source` to our SW IntrusionDetectionSystem.
+Consider InboundTrafficData as an audit collection where client IPs are added as they make requests to the server.
+As a new record is added, an event is emitted by collection streams associated with this collection. This event is consumed by the `source` InboundTraffic.
+
+Now it is quite possible that your application, `IntrusionDetectionSystem` may get burst of events having the same IP, and you need to gauge  
+this traffic in some way. Maybe 50 requests in a couple of seconds from the same IP can be considered suspicious traffic.
+Windows store events as and when they arrive and automatically expire/clean them based on the given window constraint.
+
+In our IntrusionDetectionSystem application, we have associated a `sliding window` of 2 seconds on incoming events. The window will hold all the events that appear
```
Comment (diana-macrometa) [inline docs/cep/tutorials/create-stream-app.md]:
```suggestion
In our IntrusionDetectionSystem application, we have associated a `sliding window` of two seconds on incoming events. The window will hold all the events that appear
```

File: docs/cep/tutorials/create-stream-app.md [hunk 4]
```diff
@@ -286,6 +286,73 @@ The stream worker will use the Macrometa collections with the default query para
...
+within 2 seconds. Further Stream QL is grouping all the events by IP and if the count on any of these groups is more than 50, i.e., more than 50 events occurred
```
Comment (diana-macrometa) [inline docs/cep/tutorials/create-stream-app.md]:
```suggestion
within two seconds. Further Stream QL is grouping all the events by IP and if the count on any of these groups is more than 50, i.e., more than 50 events occurred
```

File: docs/cep/tutorials/create-stream-app.md [hunk 5]
```diff
@@ -286,6 +286,73 @@ The stream worker will use the Macrometa collections with the default query para
...
+from the same IP within 2 seconds, we can assume this is suspicious traffic.
```
Comment (diana-macrometa) [inline docs/cep/tutorials/create-stream-app.md]:
```suggestion
from the same IP within two seconds, then we can assume this is suspicious traffic. 
```

File: docs/cep/tutorials/create-stream-app.md [hunk 6]
```diff
@@ -286,6 +286,73 @@ The stream worker will use the Macrometa collections with the default query para
...
+In our IntrusionDetectionSystem application, we have associated a `sliding window` of 2 seconds on incoming events. The window will hold all the events that appear
+within 2 seconds. Further Stream QL is grouping all the events by IP and if the count on any of these groups is more than 50, i.e., more than 50 events occurred
+from the same IP within 2 seconds, we can assume this is suspicious traffic. 
+
+
+In addition to the window, we have reference data for our application, i.e., `Table`.  A table is a stored collection of events, and its schema is defined via the
+table definition. Unlike `stream`, `table` is stateful and supports interactive queries for the state of the stored events. In our case `SuspiciousIPTable` stores
+pre-defined suspicious IPs which we are using to verify if it matches IPs in suspicious traffic. 
+
+If they match, we raise an alarm by adding a event to a `sink`. Sink is an abstraction for an egress event, which serves as an output for the Stream Worker application,
```
Comment (diana-macrometa) [inline docs/cep/tutorials/create-stream-app.md]:
```suggestion
If they match, then we raise an alarm by adding an event to a `sink`. Sink is an abstraction for an egress event, which serves as an output for the stream worker,
```

File: docs/cep/tutorials/create-stream-app.md [hunk 7]
```diff
@@ -286,6 +286,73 @@ The stream worker will use the Macrometa collections with the default query para
...
+convert them to various data formats, and publish them to a stream or external endpoints like HTTP/REST ,database, log, JMS, email or as in this case, a Macrometa stream.
```
Comment (diana-macrometa) [inline docs/cep/tutorials/create-stream-app.md]:
```suggestion
convert them to various data formats, and publish them to a stream or external endpoints like HTTP/REST ,database, log, JMS, email, or as in this case, a Macrometa stream.
```

File: docs/cep/tutorials/create-stream-app.md [hunk 8]
```diff
@@ -286,6 +286,73 @@ The stream worker will use the Macrometa collections with the default query para
...
+
+INSERT INTO Alarm
+SELECT ip, count(ip) as requestCount, currentTimeMillis() as incidentTime
+FROM InboundTraffic WINDOW SLIDING_TIME(2 sec) as IT
+JOIN SuspiciousIPTable as SIP
+ON  IT.ip == SIP.blocked_ip
+group by ip
+having requestCount > 50 ;
+```
+
```
Comment (diana-macrometa) [inline docs/cep/tutorials/create-stream-app.md]: Only one empty line between the example and the next heading.

**Review by diana-macrometa — CHANGES_REQUESTED** [review]

File: docs/cep/tutorials/create-stream-app.md (hunk 1 above)
Comment (jamesm-macrometa) [inline docs/cep/tutorials/create-stream-app.md]:
```suggestion
In the following application, a Macrometa document-based collection plays role of `source` to our SW IntrusionDetectionSystem.
```

File: docs/cep/tutorials/create-stream-app.md [hunk 9]
```diff
@@ -286,6 +286,73 @@ The stream worker will use the Macrometa collections with the default query para
...
+## Source, Sink, Table and Window: Understanding with the help of a Stream Worker example
+
+
+In the application below, a Macrometa document based collection plays role of `source` to our SW IntrusionDetectionSystem.
+Consider InboundTrafficData as an audit collection where client IPs are added as they make requests to the server.
+As a new record is added, an event is emitted by collection streams associated with this collection. This event is consumed by the `source` InboundTraffic.
+
+Now it is quite possible that your application, `IntrusionDetectionSystem` may get burst of events having the same IP, and you need to gauge  
+this traffic in some way. Maybe 50 requests in a couple of seconds from the same IP can be considered suspicious traffic.
```
Comment (jamesm-macrometa) [inline docs/cep/tutorials/create-stream-app.md]:
```suggestion
```

Moved content to previous line

File: docs/cep/tutorials/create-stream-app.md (hunk 2 above)
Comment (jamesm-macrometa) [inline docs/cep/tutorials/create-stream-app.md]:
```suggestion
Your application, `IntrusionDetectionSystem`, might get burst of events from IP, and you need to gauge this traffic. For example, 50 requests in a couple of seconds from the same IP can be considered suspicious traffic.
```

File: docs/cep/tutorials/create-stream-app.md [hunk 10]
```diff
@@ -286,6 +286,73 @@ The stream worker will use the Macrometa collections with the default query para
...
+Windows store events as and when they arrive and automatically expire/clean them based on the given window constraint.
+
+In our IntrusionDetectionSystem application, we have associated a `sliding window` of 2 seconds on incoming events. The window will hold all the events that appear
+within 2 seconds. Further Stream QL is grouping all the events by IP and if the count on any of these groups is more than 50, i.e., more than 50 events occurred
+from the same IP within 2 seconds, we can assume this is suspicious traffic. 
+
+
+In addition to the window, we have reference data for our application, i.e., `Table`.  A table is a stored collection of events, and its schema is defined via the
```
Comment (jamesm-macrometa) [inline docs/cep/tutorials/create-stream-app.md]:
```suggestion
In addition to the window, we have a `Table` which serves reference data for our application.  A table is a stored collection of events, and its schema is defined via the table definition. Unlike `stream`, `table` is stateful and supports interactive queries for the state of the stored events. In our case `SuspiciousIPTable` stores pre-defined suspicious IPs which we are using to verify if it matches IPs in suspicious traffic.
```

Is it correct to say that a table "serves" reference data?  Or should we rephrase this?

File: docs/cep/tutorials/create-stream-app.md [hunk 11]
```diff
@@ -286,6 +286,73 @@ The stream worker will use the Macrometa collections with the default query para
...
+table definition. Unlike `stream`, `table` is stateful and supports interactive queries for the state of the stored events. In our case `SuspiciousIPTable` stores
```
Comment (jamesm-macrometa) [inline docs/cep/tutorials/create-stream-app.md]:
```suggestion
```

Moved content to previous line

File: docs/cep/tutorials/create-stream-app.md [hunk 12]
```diff
@@ -286,6 +286,73 @@ The stream worker will use the Macrometa collections with the default query para
...
+pre-defined suspicious IPs which we are using to verify if it matches IPs in suspicious traffic.
```
Comment (jamesm-macrometa) [inline docs/cep/tutorials/create-stream-app.md]:
```suggestion
```

Moved content to previous line

**Review by jamesm-macrometa — DISMISSED** [review]
A few minor changes requested (some overlap with Diana's suggestions)

**Review by diana-macrometa — APPROVED** [review]

**Comment by diana-macrometa** [comment]
@ashish-macrometa, please review the remaining suggestions.

**Comment by diana-macrometa** [comment]
@ashish-macrometa or @StoyanTc, please finalize this topic. Respond to the suggestions and resolve conflicts.

**Pull request merged**